### PR TITLE
Autodiff - Part 1

### PR DESCRIPTION
**Pull request by NeuralCoder3**
#88 should be merged first for a cleaner transition phase.

Basic AD functionality in the new dialect form.

### Top Level
Dialect boilerplate and main definitions.
- `rules.rs`: no usage, formulation of autodiff as rewrite system,
- `autodiff.h`: axiom operations
- `normalizers.cpp`: folds sum into addition chains, simplifies addition (0+a=a), resolve congruence of additions (addition of tuples => tuple of additions), type transformation redirect
- `autodiff.thorin`: axiom declarations, currently also differentiated versions of the axioms

### Passes
Outline and headers of the passes.
- `external-cleanup`: removes differentiated axioms
- `zero cleanup`: resolves remaining zeros to bot
- `eval`: main pass, declares all important methods, memorization of rewrites, lookup tables

### Auxiliary
The important code.
- `aux`: general methods split into thorin methods and autodiff methods
  ad-methods: identity pullback, zero pullback, type transformations, zero generation, sum generation
- `rewrite_toplevel`: entry point
  generates the new lambdas, association of var (arguments), map initialization
  uses augment from rewrite_inner, corresponds to the D transformation in the paper
- `rewrite_inner`:
  core of ad, D_A transformation in the paper
  implements the augment methods
  recurses through expressions, performs modular substitution
  simple cases: var, extract, tuple, lit, pack
  interesting cases: lam, app, axiom

## Tasks

TODOs before merging:
- [x] finish this description
- [x] go through tests
  - [x] second order
  - [x] general
  - [x] direct style related
  - [x] ad 
  - [x] add, sum tests
- [x] cleanup
  - [x] includes
  - [x] empty line at bottom
  - [x] dead code
  - [x] commented code
- [x] update comments

Missing features that are added in upcoming pull requests:
- handling of eta expanded and then simplified function
- ptr axioms (load, store, alloc)
- matrix axioms 
- array axioms (lea)
- reals

### CODE REVIEW

File: lit/autodiff/multiply_autodiff_eval.thorin.eval [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    _538120#1:(.Idx 2) (1:(.Idx 4294967296), comp_tup_pb__cont_538129)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: _[[appId:[_0-9]*]]: ⊥:★ = return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296));
+// CHECK-DAG: _[[appId]]
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[retAppId:[_0-9]*]]: ⊥:★ = return_[[returnId]] _[[returnEtaVarId]];
+// CHECK-DAG: return_[[retAppId]]
```
Comment (leissa) [inline lit/autodiff/multiply_autodiff_eval.thorin.eval:70]: see above

File: lit/autodiff/multiply_autodiff_eval2.thorin.eval [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    .let _488656: (.Idx 4294967296) = %core.wrap.add (0:.Nat, 4294967296:.Nat) (_488382, _488651);
+    return_538143_488022 (mem_538158_488043, _488656)
+};
+
+// .cn .extern main __538139_488027_484299::[mem_538158_488043_484316: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_538143_488031_484303: .Cn [%mem.M, (.Idx 4294967296)]] = {
+//     .cn return_538143_488031_484294 _484304: [%mem.M, (.Idx 4294967296)] = {
+//         return_538143_488031_484303 _484304
+//     };
+//     return_538143_488031_484294 (mem_538158_488043_484316, 8402:(.Idx 4294967296))
+// };
```
Comment (leissa) [inline lit/autodiff/multiply_autodiff_eval2.thorin.eval:66]: newline

File: lit/autodiff/multiply_autodiff_eval2.thorin.eval [hunk 3]
```diff
@@ -0,0 +1,66 @@
+// RUN: rm -f %t.ll ; \
+// RUN: %thorin -d autodiff %s --output-ll %t.ll --output-thorin - | FileCheck %s
+
+// a call to a autodiff style function
+// ./build/bin/thorin -d direct -d debug -d autodiff ./lit/autodiff/multiply_autodiff_eval2.thorin --output-thorin - -VVVV
```
Comment (leissa) [inline lit/autodiff/multiply_autodiff_eval2.thorin.eval:5]: see above

File: lit/autodiff/pow_autodiff.thorin.disabled [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    // .let c = f (42:I32);
+    // return (mem, c)
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: _[[appId:[_0-9]*]]: ⊥:★ = return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296));
+// CHECK-DAG: _[[appId]]
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[retAppId:[_0-9]*]]: ⊥:★ = return_[[returnId]] _[[returnEtaVarId]];
+// CHECK-DAG: return_[[retAppId]]
```
Comment (leissa) [inline lit/autodiff/pow_autodiff.thorin.disabled:80]: see above

File: lit/autodiff/second/double_diff.thorin.disabled [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    .let c = (42:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/second/double_diff.thorin.disabled:51]: see above

File: lit/autodiff/second/snd_order.thorin.disabled [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    .let c = (42:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/second/snd_order.thorin.disabled:60]: see above

File: lit/autodiff/second/snd_order_man.thorin [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    .let c = (42:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 8702:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/second/snd_order_man.thorin:57]: see above

File: lit/autodiff/simple_autodiff.thorin [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    .let c = (43:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 2:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/simple_autodiff.thorin:42]: see above

File: lit/autodiff/square.thorin [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    .let c = (42:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 1764084:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/square.thorin:42]: see above

File: lit/autodiff/square_autodiff.thorin [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    .let c = (42:I32);
+    f_diff_cast (c,ret_cont)
+};
+
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 176484:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/autodiff/square_autodiff.thorin:44]: see above

File: lit/direct/Issue.txt [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ds_dependent.thorin
+
+same as ds2cps_ax_cps2ds_dependent.thorin but only after optimization
+
+
+
+
+ds2cps_ax_cps2ds_dependent_short.thorin
+
+:4294967295: error: 'Π n_173817: .Nat → .infer (<nullptr>)' used as a type but is in fact a term
```
Comment (leissa) [inline lit/direct/Issue.txt:19]: Can we remove this file and instead open a real issue?

File: lit/direct/complex1.thorin.eval [hunk 12]
```diff
@@ -0,0 +1,63 @@
+.import core;
+.import debug;
+.import direct;
+.import mem;
+// .lam .extern internal_diff_core_icmp_xYgLE __690955: .Nat → .Cn [«2:.Nat; .Idx __690955», .Cn [(.Idx 2), .Cn [(.Idx 2), .Cn «2:.Nat; .Idx __690955»]]] = {
```
Comment (leissa) [inline lit/direct/complex1.thorin.eval:5]: what's the deal with this commented code section?

File: lit/direct/complex1.thorin.eval [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    .cn comp_tup_pb__cont_691717 _691882::[_691883: (.Idx 4294967296), _691945: (.Idx 4294967296)] = {
+        .let _691796: (.Idx 4294967296) = %core.wrap.mul (0:.Nat, 4294967296:.Nat) (100:(.Idx 4294967296), _691708#0:(.Idx 2));
+        .let _691884: (.Idx 4294967296) = %direct.cps2ds_dep ((.Idx 4294967296), Uf_691827) zero_pb_691875 _691883;
+        .let _691946: (.Idx 4294967296) = %direct.cps2ds_dep ((.Idx 4294967296), Uf_691889) id_pb_691937 _691945;
+        .let _691951: (.Idx 4294967296) = %core.wrap.add (0:.Nat, 4294967296:.Nat) (_691884, _691946);
+        .let _691956: (.Idx 4294967296) = %core.wrap.add (0:.Nat, 4294967296:.Nat) (_691796, _691951);
+        return_691730 (mem_691743, _691956)
+    };
+    _691708#1:(.Idx 2) (1:(.Idx 4294967296), comp_tup_pb__cont_691717)
+};
```
Comment (leissa) [inline lit/direct/complex1.thorin.eval:63]: newline

File: lit/direct/ds2cps_ax_cps2ds.thorin [hunk 14]
```diff
@@ -20,8 +22,8 @@
     return (mem, c)
 };
 
-// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]+]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]+]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
-// CHECK-DAG: return_[[returnEtaId:[_0-9]+]] (mem_[[memId]], 42:(.Idx 4294967296))
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296))
 
 // CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
 // CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/direct/ds2cps_ax_cps2ds.thorin:29]: see above

File: lit/direct/ds2cps_ax_cps2ds_dependent.thorin.disabled [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        // )
+        c;
+    return (mem, c_cast)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]+]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_[[returnId:[_0-9]+]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]+]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/direct/ds2cps_ax_cps2ds_dependent.thorin.disabled:49]: see above

File: lit/direct/ds2cps_ax_cps2ds_dependent2.thorin [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    .let n = 4294967296;
+    .let c = %direct.cps2ds_dep (I32,U) (f (n, 0)) (5:(.Idx n));
+    return (mem, c)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]+]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]+]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]+]] (mem_[[memId]], 47:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/direct/ds2cps_ax_cps2ds_dependent2.thorin:38]: see above

File: lit/direct/ds2cps_ax_cps2ds_dependent_short.thorin.disabled [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    .let g = %direct.cps2ds_dep (.Nat,lam_Idx) f;
+    .let c = g (4294967296);
+    return (mem, c)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]+]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_[[returnId:[_0-9]+]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]+]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/direct/ds2cps_ax_cps2ds_dependent_short.thorin.disabled:29]: see above

File: lit/direct/ds_dependent.thorin.disabled [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    .let n = 4294967296;
+    .let c = f n;
+    return (mem, c)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]+]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0:.Nat), 0:.Nat), return_[[returnId:[_0-9]+]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]+]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/direct/ds_dependent.thorin.disabled:32]: see above

File: lit/tool/filter.thorin [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    .let c = (42:I32);
+    f (c,cont1)
+};
+
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 168:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/tool/filter.thorin:39]: see above

File: lit/tool/force.thorin [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+.cn .extern main [mem : %mem.M, argc : I32, argv : %mem.Ptr (%mem.Ptr (.Idx 256, 0:.Nat), 0:.Nat), return : .Cn [%mem.M, I32]] = {
+    .let c = (42:(%tool.force_type(U 4294967296)));
+    return (mem, c)
+};
+
+// CHECK-DAG: .cn .extern main _{{[0-9_]+}}::[mem_[[memId:[_0-9]*]]: %mem.M, (.Idx 4294967296), %mem.Ptr (%mem.Ptr ((.Idx 256), 0), 0), return_[[returnId:[_0-9]*]]: .Cn [%mem.M, (.Idx 4294967296)]] = {
+// CHECK-DAG: return_[[returnEtaId:[_0-9]*]] (mem_[[memId]], 42:(.Idx 4294967296))
+
+// CHECK-DAG: return_[[returnEtaId]] _[[returnEtaVarId:[0-9_]+]]: [%mem.M, (.Idx 4294967296)] = {
+// CHECK-DAG: return_[[returnId]] _[[returnEtaVarId]]
```
Comment (leissa) [inline lit/tool/force.thorin:24]: see above

File: thorin/analyses/deptree.cpp [hunk 21]
```diff
@@ -26,9 +26,21 @@ VarSet DepTree::run(Def* nom) {
     auto parent = root_.get();
     for (auto var : result) {
         auto n = nom2node_[var->nom()].get();
+        if (!n) {
+            const_cast<World&>(world_).ELOG("var {} used before nom {} discovered, old var still around?", var,
+                                            var->nom());
+            const_cast<World&>(world_).ELOG("var {} : {}", var, var->type());
+            const_cast<World&>(world_).ELOG("var nom {} : {}", var->nom(), var->nom()->type());
```
Comment (leissa) [inline thorin/analyses/deptree.cpp:33]: I think we should remove the `const_cast` and instead declare `World::State::POD::log` as `mutable`.

File: thorin/pass/pipelinebuilder.h [hunk 22]
```diff
@@ -33,6 +33,8 @@ class PipelineBuilder {
     std::vector<int> passes();
 
 private:
+    // std::map<int, std::vector<std::function<void(PassMan&)>>> phase_extensions_;
+    // std::vector<std::function<void(PassMan&)>> codegen_prep_phase_extensions_;
```
Comment (leissa) [inline thorin/pass/pipelinebuilder.h:37]: remove

File: thorin/world.h [hunk 23]
```diff
@@ -145,8 +145,10 @@ class World {
     bool empty() { return move_.externals.empty(); }
     void make_external(Def* def) {
         assert(!def->name().empty());
-        auto [_, ins] = move_.externals.emplace(def->name(), def);
-        assert(ins);
+        auto name = def->name();
+        // assert(!move_.externals.contains(name));
+        auto [i, ins] = move_.externals.emplace(def->name(), def);
+        // assert((ins || (def == i->second)) && "make sure the external isn't inserted twice");
```
Comment (leissa) [inline thorin/world.h:151]: we had the discussion in Discord. Can we enable the assertion again?

File: thorin/world.h [hunk 24]
```diff
@@ -145,8 +145,10 @@ class World {
     bool empty() { return move_.externals.empty(); }
     void make_external(Def* def) {
         assert(!def->name().empty());
-        auto [_, ins] = move_.externals.emplace(def->name(), def);
-        assert(ins);
+        auto name = def->name();
+        // assert(!move_.externals.contains(name));
```
Comment (leissa) [inline thorin/world.h:149]: If the 2nd assert is enabled again, this one here is superfluous.

**Review by leissa — CHANGES_REQUESTED** [review]
Fantastic work and most of the stuff looks pretty good to more. Just some usual nitpicking by me :)

One general remarks regarding the Doxygen docs:
* Can you please make sure that you are using complete sentences that start with a capital letter and end with a full stop?
  If it's really just a small note, that you can keep it like this, but often you are using complete sentences which are not formatted as sentences.
* Please use one sentence per line.
* Please format your Doxygen comments with markdown commands. Some docs look pretty broken as the formatting is lacking. E.e. put code examples in fenced code blocks, use item lists where appropriate etc.

File: .vscode/launch.json [hunk 25]
```diff
@@ -0,0 +1,51 @@
+{
+    // Verwendet IntelliSense zum Ermitteln möglicher Attribute.
+    // Zeigen Sie auf vorhandene Attribute, um die zugehörigen Beschreibungen anzuzeigen.
+    // Weitere Informationen finden Sie unter https://go.microsoft.com/fwlink/?linkid=830387
```
[the remark this answers is not in the document]
Comment (NeuralCoder3) [reply .vscode/launch.json:4]: Yes, we should polish it a bit and remove the german comment.
I think in general, a launch config for a popular IDE (like VSCode) for debugging with dialects might be helpful.
(And I would not need to add it manually every time I clone the repo)

File: thorin/analyses/deptree.cpp (hunk 21 above)
Comment (leissa) [reply thorin/analyses/deptree.cpp:33]: See my latest commit on `master`

File: dialects/tool/tool.thorin [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+///
+/// ### %tool.force
+/// 
+/// Forces beta reduction
+.ax %tool.force_type: * -> *, normalize_force_type;
+///
+/// ### %tool.set_filter
+/// 
+// .ax %tool.set_filter: Π [T: *] -> [.Idx 2,T] -> T, normalize_set_filter;
+.ax %tool.set_filter: [.Idx 2,⊥:★] -> ⊥:★, normalize_set_filter;
```
[the remark this answers is not in the document]
Comment (NeuralCoder3) [reply dialects/tool/tool.thorin:25]: This issue only concerns the tool dialect and tangentially the direct dialect.
But it does not affect the ad dialect. I move the discussion of this issue to the corresponding pull request.

**Comment by leissa** [comment]
Does somebody know what the heck is wrong with this failing Windows test?

File: .vscode/launch.json (hunk 25 above)
[the remark this answers is not in the document]
Comment (leissa) [reply .vscode/launch.json:4]: Yes, exactly. Then comes someone along who uses KDevelop and puts kdevelop project files into our repo. Next, we have an eclipse guy putting an eclipse project into the repo. Next is Netbeans. Then comes Codeblocks, etc ...

File: lit/autodiff/general/add_tuple.thorin [hunk 27]
```diff
@@ -0,0 +1,24 @@
+// RUN: rm -f %t.ll ; \
+// RUN: %thorin -d autodiff %s --output-ll %t.ll --output-thorin - | FileCheck %s
+
+// ./build/bin/thorin -d autodiff ./lit/autodiff/general/add_tuple.thorin --output-thorin -
```
[the remark this answers is not in the document]
Comment (NeuralCoder3) [reply lit/autodiff/general/add_tuple.thorin:4]: This comment seems different.
What is the suggested improvement here?

File: lit/autodiff/multiply_autodiff_cond_eval3_2.thorin.eval [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+        .cn comp_tup_pb__cont_635804_709379 _635869_709463::[(.Idx 4294967296), _635933_709464: (.Idx 4294967296)] = {
+            .let _709450: (.Idx 4294967296) = %core.wrap.mul (0:.Nat, 4294967296:.Nat) (100:(.Idx 4294967296), _635972_709439);
+            .let _709469: (.Idx 4294967296) = %core.wrap.add (0:.Nat, 4294967296:.Nat) (_709450, _635933_709464);
+            return_635714_709385 (mem_635729_709403, _709469)
+        };
+            _635800_709365 (1:(.Idx 4294967296), comp_tup_pb__cont_635804_709379)
+    };
+    inner_mul_deriv_cps_635298_709183 aug_thrice_cont_635794_709328
+};
+/*
```
[the remark this answers is not in the document]
Comment (NeuralCoder3) [reply lit/autodiff/multiply_autodiff_cond_eval3_2.thorin.eval:34]: This is the final evaluation result (after three loops through thorin -> fix output -> run -> repeat). I will move it to a new file to make it clearer.

File: lit/direct/complex1.thorin.eval (hunk 12 above)
Comment (NeuralCoder3) [reply lit/direct/complex1.thorin.eval:5]: It was the remaining externals as the code failed/was aborted before the external sweep pass ran.

File: lit/direct/ds2cps_ax_cps2ds.thorin (hunk 14 above)
Comment (NeuralCoder3) [reply lit/direct/ds2cps_ax_cps2ds.thorin:29]: transitively resolved

File: .vscode/launch.json (hunk 25 above)
[the remark this answers is not in the document]
Comment (NeuralCoder3) [reply .vscode/launch.json:4]: > Yes, exactly. Then comes someone along who uses KDevelop and puts kdevelop project files into our repo. Next, we have an eclipse guy putting an eclipse project into the repo. Next is Netbeans. Then comes Codeblocks, etc ...

Okay. I see the point.

File: lit/autodiff/general/add_tuple.thorin (hunk 27 above)
[the remark this answers is not in the document]
Comment (leissa) [reply lit/autodiff/general/add_tuple.thorin:4]: What is the gain in having this here? If your lit test fails you see the exact command line that you just need to copy & paste.
The problem I have with such comments is that it's usually just a matter of time till the `RUN` command gets updated and the comment command is forgotten. Then, you are in the situation where the command in the comment doesn't match the `RUN` command and the guessing starts: Who is right?

File: thorin/world.h (hunk 23 above)
Comment (NeuralCoder3) [reply thorin/world.h:151]: We discussed that it is a problem and should not happen. But we did not resolve it.

Comment (NeuralCoder3) [reply thorin/world.h:151]: Moved to #99

**Review by leissa — APPROVED** [review]

**Pull request merged**